### PR TITLE
Properly serialize remote query in ReindexRequest

### LINKED ISSUE #43406: Reindex from remote query can be serialized incorrectly
Currently when you reindex from remote, the query is stored in a serialized format with a content type. When `ReindexRequest` is serialized, it attempts to serialize the query using the content-type that xcontent is being serialized to. This might be different and can throw an exception.

It does look like the `RemoteInfo` always stores the query as JSON so ensuring that it is serialized as JSON in `ReindexRequest` might be the fix.

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit modifies the `RemoteInfo` to clarify that a search query
must always be serialized as JSON. Additionally, it adds an assertion
to ensure that this is the case. This fixes #43406.

Additionally, this PR implements `AbstractXContentTestCase` for the
reindex request. This is related to #43456.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java [hunk 1]
```diff
@@ -41,11 +48,97 @@
...
+    @Override
+    protected void assertEqualInstances(ReindexRequest expectedInstance, ReindexRequest newInstance) {
+        assertNotSame(newInstance, expectedInstance);
+        assertArrayEquals(expectedInstance.getSearchRequest().indices(), newInstance.getSearchRequest().indices());
+        assertEquals(expectedInstance.getSearchRequest(), newInstance.getSearchRequest());
+        assertEquals(expectedInstance.getMaxDocs(), newInstance.getMaxDocs());
+        assertEquals(expectedInstance.getSlices(), newInstance.getSlices());
+        assertEquals(expectedInstance.isAbortOnVersionConflict(), newInstance.isAbortOnVersionConflict());
+        assertEquals(expectedInstance.getRemoteInfo(), newInstance.getRemoteInfo());
+    }
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java]: I think we should assert the remaining fields too, like dest-op-type, dest-pipeline and more.

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteWhitelistTests.java [hunk 2]
```diff
@@ -37,6 +39,9 @@
  * Tests the reindex-from-remote whitelist of remotes.
  */
 public class ReindexFromRemoteWhitelistTests extends ESTestCase {
+
+    private final BytesReference matchAll = new BytesArray(new MatchAllQueryBuilder().toString());
```
Comment (henningandersen) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFromRemoteWhitelistTests.java]: I would prefer `new BytesArray("{ \"foo\" : \"bar\" }"`, makes it more explicit what we rely on (JSON and not query-validity).

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexSourceTargetValidationTests.java [hunk 3]
```diff
@@ -61,6 +63,8 @@
     private static final AutoCreateIndex AUTO_CREATE_INDEX = new AutoCreateIndex(Settings.EMPTY,
             new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS), INDEX_NAME_EXPRESSION_RESOLVER);
 
+    private final BytesReference matchAll = new BytesArray(new MatchAllQueryBuilder().toString());
```
Comment (henningandersen) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexSourceTargetValidationTests.java]: I would prefer `new BytesArray("{ \"foo\" : \"bar\" }"`, makes it more explicit what we rely on (JSON and not query-validity).

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteInfoTests.java [hunk 4]
```diff
@@ -19,28 +19,28 @@
...
+import org.elasticsearch.index.reindex.RemoteInfo;
 import org.elasticsearch.test.ESTestCase;
 
 import static java.util.Collections.emptyMap;
 
 public class RemoteInfoTests extends ESTestCase {
     private RemoteInfo newRemoteInfo(String scheme, String prefixPath, String username, String password) {
-        return new RemoteInfo(scheme, "testhost", 12344, prefixPath, new BytesArray("testquery"), username, password, emptyMap(),
-                RemoteInfo.DEFAULT_SOCKET_TIMEOUT, RemoteInfo.DEFAULT_CONNECT_TIMEOUT);
+        return new RemoteInfo(scheme, "testhost", 12344, prefixPath,new BytesArray("{\"match_all\":{}}"), username, password, emptyMap(),
```
Comment (henningandersen) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteInfoTests.java]: I would prefer `new BytesArray("{ \"foo\" : \"bar\" }"`, makes it more explicit what we rely on (JSON and not query-validity).

File: server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java [hunk 5]
```diff
@@ -41,11 +48,97 @@
  */
 public class ReindexRequestTests extends AbstractBulkByScrollRequestTestCase<ReindexRequest> {
 
+    private final String matchAll = "{\"match_all\":{}}";
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java]: I would prefer `new BytesArray("{ \"foo\" : \"bar\" }"`, makes it more explicit what we rely on (JSON and not query-validity).

File: server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java [hunk 6]
```diff
@@ -221,6 +314,8 @@ public void testReindexFromRemoteRequestParsing() throws IOException {
         }
     }
 
+
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/reindex/ReindexRequestTests.java]: nit: extra newlines

**Review by henningandersen — APPROVED** [review]
LGTM.

I left a few smaller comments to address but think there is no need for another review round.

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteInfoTests.java (hunk 4 above)
Comment (tbrooks8) [reply modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteInfoTests.java]: `"{\"match_all\":{}}"` is JSON?

Comment (tbrooks8) [reply modules/reindex/src/test/java/org/elasticsearch/index/reindex/remote/RemoteInfoTests.java]: Oh, you just just want a nonsense query that is very clearly JSON I think?

**Pull request merged**